File: benchmarks/call-closure-poly4/Bench.java
public class Bench {
    @FunctionalInterface
    interface IntFunc {
        int apply(int x);
    }

    static IntFunc makeAdder(int base) {
        return x -> x + base;
    }

    public static void main(String[] args) {
        int iterations = 20_000_000;
        IntFunc f0 = makeAdder(1);
        IntFunc f1 = makeAdder(3);
        IntFunc f2 = makeAdder(5);
        IntFunc f3 = makeAdder(7);
        long acc = 0;

        for (int i = 0; i < iterations; i++) {
            int x = i & 1023;
            int slot = i & 3;
            if (slot == 0) {
                acc += f0.apply(x);
            } else if (slot == 1) {
                acc += f1.apply(x);
            } else if (slot == 2) {
                acc += f2.apply(x);
            } else {
                acc += f3.apply(x);
            }
        }

        System.out.println(acc);
    }
}

File: benchmarks/call-closure-poly4/bench.go
package main

import "fmt"

func makeAdder(base int) func(int) int {
	return func(x int) int {
		return x + base
	}
}

func main() {
	iterations := 20000000
	f0 := makeAdder(1)
	f1 := makeAdder(3)
	f2 := makeAdder(5)
	f3 := makeAdder(7)
	acc := 0

	for i := 0; i < iterations; i++ {
		x := i & 1023
		slot := i & 3
		if slot == 0 {
			acc += f0(x)
		} else if slot == 1 {
			acc += f1(x)
		} else if slot == 2 {
			acc += f2(x)
		} else {
			acc += f3(x)
		}
	}

	fmt.Println(acc)
}
